src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/naxxramas/boss_four_horsemen.cpp

```diff
@@ -142,6 +142,17 @@ struct boss_horsmenAI : public ScriptedAI
             m_instance->SetData(TYPE_FOUR_HORSEMEN, SPECIAL);
     }
 
+    void EnterEvadeMode() override
+    {
+        ScriptedAI::EnterEvadeMode();
+
+        std::vector<uint32> entries = {NPC_BLAUMEUX, NPC_MOGRAINE, NPC_ZELIEK, NPC_THANE};
+        for (uint32 entry : entries)
+            if (Creature* creature = m_instance->GetSingleCreatureFromStorage(entry) )
+                if (creature->IsInCombat())
+                    creature->AI()->EnterEvadeMode();
+    }
+
```

Comment:
I agree that Horsemen aggro and evade all at once, but it should not be possible to have one of the Horsemen enter evade mode (*i.e.* having no target) with one of the other still having one, so I'm puzzled by this code.

If it still still needed, I suggest using instead the [encounter failure code](https://github.com/cmangos/mangos-tbc/blob/1acd44e15f3ee03b3394a41c8fce15a26d53a445/src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/naxxramas/instance_naxxramas.cpp#L529) in `instance_naxxramas::SetData()`

Reply:
1)
Target going outside of the room/beyond gates does not make the target unreachable, because boss/npc pathing *always* ignores toggleable GO collisions, for them the gate does not exist, and they don't know the room boundary without the leash value
2)
Doing the horseman fight like this might be useless and I haven't tried, but theoretically this should be possible: A raid stands in front of the 4HM. 3 players are deliberately set outside of the Arena, the rest of the raid is just inside the arena but out of aggro range. The encounter is started and 3 hunters misdirect 3 of the horsemen to the 3 players waiting outside of the leashing zone, the remaining horseman is tanked at the room entrance. 3 Horsemen run outside of leash range so they reset, one horseman is still active tho, so the encounter hasn't failed, 3 horsemen are now idle on the platform, while 1 is being killed solo. Rinse and repeat until all 4 horsemen were individually killed, without the encounter having technically failed.
(This is purely hypothetical and it might not work like that for a myriad of reasons)
The point is mostly to ensure that once one horseman evades for whatever reason, so do the others.

Reply:
Two things. Cala is right, the code for this should only be in m_instance->SetData(TYPE_FOUR_HORSEMEN, FAIL);

However, the m_instance->SetData(TYPE_FOUR_HORSEMEN, FAIL); should be on EVADE not on home. Our scripts have that largely wrong. Blizzard despawns/resets encounters on evade not on home arrival.

Reply:
Should I overhaul this in https://github.com/cmangos/mangos-tbc/blob/master/src/game/AI/ScriptDevAI/scripts/eastern_kingdoms/naxxramas/instance_naxxramas.cpp#L529 then?

```cpp
                // Despawn spirits
                if (Creature* spirit = GetSingleCreatureFromStorage(NPC_SPIRIT_OF_BLAUMEUX))
                    spirit->ForcedDespawn();
                if (Creature* spirit = GetSingleCreatureFromStorage(NPC_SPIRIT_OF_MOGRAINE))
                    spirit->ForcedDespawn();
                if (Creature* spirit = GetSingleCreatureFromStorage(NPC_SPIRIT_OF_KORTHAZZ))
                    spirit->ForcedDespawn();
                if (Creature* spirit = GetSingleCreatureFromStorage(NPC_SPIRIT_OF_ZELIREK))
                    spirit->ForcedDespawn();
```
is currently only called when the encounter is in state `DONE` do the spirits clean themselves in other fail-states?

Otherwise I'd refactor that part to include both failed and done encounters, as well as include the evade code in the failed encounter state

Reply:
I agree with @killerwife : as far I know encounters reset on evade, not reached home. It is adds respawning that happens on reached home (at least the few cases I have in mind, probably because they are done through spells).

@insunaa Yes, you should overhaul SetData so the spirits also clean themselves on `FAIL`, please.

Reply:
@cala made the changes, but now I have a problem where the 4HM re-aggro when slightly outside of the Detection range
https://www.youtube.com/watch?v=mAGdesoOSkQ

Reply:
A boxed leash is perfectly valid when its warranted, do not shy away from it. It should only be in db if its circular and at spawn.